amp-beta-tester.php

```diff
@@ -0,0 +1,292 @@
+add_action( 'plugins_loaded', __NAMESPACE__ . '\init' );
+
+/**
+ * Force a plugin update check. This will allows us to modify the plugin update cache so we
+ * can set a custom update.
+ */
+function force_plugin_update_check() {
+	if ( wp_doing_cron() ) {
+		return;
+	}
+	delete_site_transient( 'update_plugins' );
+}
+
+/**
+ * Restore `update_plugins` transient and remove any plugin data.
+ */
+function restore_update_plugins_transient() {
+	delete_site_transient( AMP__BETA__TESTER__RELEASES__TRANSIENT );
+	delete_site_transient( 'update_plugins' );
+}
+
+/**
+ * Hook into WP.
+ *
+ * @return void
+ */
+function init() {
+	add_filter( 'plugins_api_result', __NAMESPACE__ . '\update_amp_plugin_information', 10, 3 );
+	add_filter( 'pre_set_site_transient_update_plugins', __NAMESPACE__ . '\update_amp_manifest' );
+}
+
+/**
+ * Modifies the AMP plugin manifest to point to the latest non-stable update, if it exists.
+ *
+ * @param \stdClass $updates Object containing information on plugin updates.
+ * @return \stdClass
+ */
+function update_amp_manifest( $updates ) {
+	if ( ! isset( $updates->no_update ) ) {
+		return $updates;
+	}
+
+	if ( ! get_current_amp_update_manifest() ) {
+		return $updates;
+	}
+
+	$latest_manifest = get_amp_update_manifest();
+
+	if ( ! $latest_manifest ) {
+		$current_manifest = get_amp_update_manifest( get_amp_version() );
+
+		if ( $current_manifest ) {
+			unset( $updates->response[ AMP__PLUGIN__BASENAME ] );
+			$updates->no_update[ AMP__PLUGIN__BASENAME ] = $current_manifest;
+		}
+	} else {
+		unset( $updates->no_update[ AMP__PLUGIN__BASENAME ] );
+		$updates->response[ AMP__PLUGIN__BASENAME ] = $latest_manifest;
+	}
+
+	return $updates;
+}
+
+/**
+ * Update the AMP plugin information data to reflect that of the GitHub release.
+ *
+ * @param false|object|array $value  The result object or array. Default false.
+ * @param string             $action The type of information being requested from the Plugin Installation API.
+ * @param object             $args   Plugin API arguments.
+ * @return false|object|array Updated $value, or passed-through $value on failure.
+ */
+function update_amp_plugin_information( $value, $action, $args ) {
+	if ( 'plugin_information' !== $action || 'amp' !== $args->slug ) {
+		return $value;
+	}
+
+	$amp_version = get_amp_version();
+
+	if ( ! is_pre_release( $amp_version ) ) {
+		return $value;
+	}
+
+	$amp_manifest = get_amp_update_manifest( $amp_version );
+
+	if ( ! $amp_manifest ) {
+		return $value;
+	}
+
+	$value->version = $amp_version;
+
+	$value->download_link            = $amp_manifest->package;
+	$value->versions[ $amp_version ] = $amp_manifest->package;
+
+	return $value;
+}
+
+/**
+ * Fetch AMP releases from GitHub.
+ *
+ * @return array|false
+ */
+function get_amp_github_releases() {
+	$raw_response = wp_remote_get( 'https://api.github.com/repos/ampproject/amp-wp/releases' );
+	if ( is_wp_error( $raw_response ) ) {
+		return false;
```

Comment:
The way this is combined with the transient can cause major issues if the connection is faulty or GitHub has issues.

As this returns `false` on error, the consuming code will store `false` in a transient, and in a subsequent request check if `empty( false )`, thus making a remote GitHub request on every single page request until the situation is resolved. That can easily get the server blacklisted.

Returning an empty array here, and checking for `false` in the consuming code would store the faulty response and only retry much later instead.
```suggestion
		return [];
```

Reply:
No longer the case anymore as of e2c009787c95cb7cff7e4f73bb13e32ac64f184a.